spring-cloud-scheduler-spi-test/src/main/java/org/springframework/cloud/scheduler/spi/test/AbstractIntegrationTests.java

```diff
@@ -0,0 +1,461 @@
+
+/**
+ * Contains base set of tests that are required for each implementation of
+ * Spring Cloud Scheduler to pass.
+ *
+ * @author Glenn Renfro
+ */
+@RunWith(SpringRunner.class)
+@SpringBootTest(webEnvironment = NONE)
+@ContextConfiguration(classes = AbstractIntegrationTests.Config.class)
+public abstract class AbstractIntegrationTests {
+
+	@Autowired
+	protected MavenProperties mavenProperties;
+
+	private SchedulerWrapper schedulerWrapper;
+
+	protected final Logger log = LoggerFactory.getLogger(this.getClass());
+
+	@Rule
+	public TestName name = new TestName();
+
+	@After
+	public void tearDown() {
+		List<ScheduleRequest> scheduleRequests = new ArrayList<>(schedulerWrapper.getScheduledTasks().values());
+
+		for (ScheduleRequest scheduleRequest : scheduleRequests) {
+			unschedule(scheduleRequest.getScheduleName());
+		}
+
+	}
+
+	/**
+	 * Subclasses should call this method to interact with the Scheduler under test.
+	 * Returns a wrapper around the scheduler returned by {@link #provideScheduler()}, that keeps
+	 * track of which tasks have been scheduled and unscheduled.
+	 */
+	protected Scheduler taskScheduler() {
+		return this.schedulerWrapper;
+	}
+
+	/**
+	 * To be implemented by subclasses, which should return the instance of Scheduler that needs
+	 * to be tested. Can be used if subclasses decide to add additional implementation-specific tests.
+	 */
+	protected abstract Scheduler provideScheduler();
+
+	/**
+	 * To be implemented by subclasses, which should return the commandLineArgs that
+	 * will be used for the tests.
+	 */
+	protected abstract List<String> testCommandLineArgs();
+
+	/**
+	 * To be implemented by subclasses, which should return the schedulerProperties that
+	 * will be used for the tests.
+	 */
+	protected abstract Map<String, String> testSchedulerProperties();
+
+	/**
+	 * To be implemented by subclasses, which should return the deploymentProperties that
+	 * will be used for the tests.
+	 */
+	protected abstract Map<String, String> testDeploymentProperties();
+
+	/**
+	 * To be implemented by subclasses, which should return the appProperties that
+	 * will be used for the tests.
+	 */
+	protected abstract Map<String, String> testAppProperties();
+
+	@Before
+	public void wrapScheduler() {
+		this.schedulerWrapper = new SchedulerWrapper(provideScheduler());
+	}
+
+	@Test
+	public void testSimpleSchedule() {
+		createAndVerifySchedule();
+	}
+
+	@Test
+	public void testUnschedule() {
+		ScheduleInfo scheduleInfo = createAndVerifySchedule();
+		unschedule(scheduleInfo.getScheduleName());
+	}
+
+	@Test
+	public void testDuplicateSchedule() {
+		boolean exceptionFired = false;
+		ScheduleRequest request = createScheduleRequest();
+		schedule(request, request.getDefinition());
+		ScheduleInfo scheduleInfo = new ScheduleInfo();
+		scheduleInfo.setScheduleName(request.getScheduleName());
+
+		verifySchedule(scheduleInfo);
+		try {
+			schedule(request, request.getDefinition());
```

Comment:
Personal preference here, but this feels like an unnecessary level of indirection.  As a dev, I have to go into the schedule test to figure out what it does even though it only logs a message and then does the actual scheduling.

Reply:
done